File: StructureSolver_Parallel/StructureSolver.cpp
#include "StructureSolver.h"
#include "precice/SolverInterface.hpp"

using namespace precice;
using namespace precice::constants;

int main(int argc, char** argv)
{

  std::cout << "Starting Structure Solver..." << std::endl;
  if (argc != 3) {
    std::cout << std::endl;
    std::cout << "Structure: Usage: mpiexec -np <#procs> " << argv[0] << " <configurationFileName> <N>" << std::endl;
    std::cout << std::endl;
    std::cout << "N:     Number of mesh elements, needs to be equal for fluid and structure solver." << std::endl;
    return -1;
  }

  MPI_Init(&argc, &argv);

  int domainSize, gridOffset, rank, size, chunkLength, *vertexIDs;
  double *pressure, *crossSectionLength, *grid; // Declare dataset

  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

  domainSize = atoi(argv[2]);
  if ((domainSize + 1) % size == 0) {
    chunkLength = (domainSize + 1) / size;
    gridOffset = rank * chunkLength;
  } else if (rank < (domainSize + 1) % size) {
    chunkLength = (domainSize + 1) / size + 1;
    gridOffset = rank * chunkLength;
  } else {
    chunkLength = (domainSize + 1) / size;
    gridOffset = ((domainSize + 1) % size) * ((domainSize + 1) / size + 1) + (rank - ((domainSize + 1) % size)) * (domainSize + 1) / size;
  }

  pressure = new double[chunkLength];
  crossSectionLength = new double[chunkLength];

  std::string configFileName(argv[1]);
  std::string solverName = "STRUCTURE";

  SolverInterface interface(solverName, rank, size);
  interface.configure(configFileName);

  int meshID = interface.getMeshID("Structure_Nodes");
  int pressureID = interface.getDataID("Pressure", meshID);
  int crossSectionLengthID = interface.getDataID("CrossSectionLength", meshID);

  int dimensions = interface.getDimensions();
  grid = new double[dimensions * chunkLength];
  vertexIDs = new int[chunkLength];

  for (int i = 0; i < chunkLength; i++) {
    crossSectionLength[i] = 1.0;
    pressure[i] = 0.0;
    for (int j = 0; j < dimensions; j++) {
      grid[i * dimensions + j] = j == 0 ? gridOffset + (double)i : 0.0;
    }
  }

  interface.setMeshVertices(meshID, chunkLength, grid, vertexIDs);

  interface.initialize();

  double t = 0;
  double dt = 0.01;

  if (interface.isActionRequired(actionWriteInitialData())) {
    interface.writeBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs, crossSectionLength);
    interface.fulfilledAction(actionWriteInitialData());
  }

  interface.initializeData();

  if (interface.isReadDataAvailable()) {
    interface.readBlockScalarData(pressureID, chunkLength, vertexIDs, pressure);
  }

  while (interface.isCouplingOngoing()) {
    if (interface.isActionRequired(actionWriteIterationCheckpoint())) {
      interface.fulfilledAction(actionWriteIterationCheckpoint());
    }

    structureComputeSolution(rank, size, chunkLength, pressure, crossSectionLength); // Call Solver
                                                                                     //structureDataDisplay(crossSectionLength, chunkLength);
                                                                                     //structureDataDisplay(pressure, chunkLength);

    interface.writeBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs, crossSectionLength);

    interface.advance(dt);

    interface.readBlockScalarData(pressureID, chunkLength, vertexIDs, pressure);

    if (interface.isActionRequired(actionReadIterationCheckpoint())) { // i.e. fluid not yet converged
      interface.fulfilledAction(actionReadIterationCheckpoint());
    } else {
      t += dt;
    }
  }

  delete (pressure);
  delete (crossSectionLength);
  delete (grid);

  MPI_Finalize();

  return 0;
}

File: PythonTube/StructureSolver.py
from __future__ import division, print_function

import os
import sys
import argparse
from mpi4py import MPI
import numpy as np
import configuration_file as config
 
import precice
from precice import *

print("Starting Structure Solver...")

parser = argparse.ArgumentParser()
parser.add_argument("configurationFileName", help="Name of the xml config file.", type=str)

try:
    args = parser.parse_args()
except SystemExit:
    print("")
    print("Did you forget adding the precice configuration file as an argument?")
    print("Try $python StructureSolver.py precice-config.xml")
    quit()

configFileName = args.configurationFileName
N = config.n_elem

print("N: " + str(N))

solverName = "STRUCTURE"

print("Configure preCICE...")
interface = precice.Interface(solverName, 0, 1)
interface.configure(configFileName)
print("preCICE configured...")

dimensions = interface.get_dimensions()

pressure = config.p0 * np.ones(N+1)
crossSectionLength = config.a0 * np.ones(N+1)

meshID = interface.get_mesh_id("Structure_Nodes")
crossSectionLengthID = interface.get_data_id("CrossSectionLength", meshID)
pressureID = interface.get_data_id("Pressure", meshID)

vertexIDs = np.zeros(N+1)
grid = np.zeros([dimensions, N+1])

grid[0,:] = np.linspace(0, config.L, N+1)  # x component
#grid[1,:] = np.linspace(0, config.L, N+1)  # y component, leave blank

interface.set_mesh_vertices(meshID, N+1, grid.flatten('F'), vertexIDs)

t = 0

print("Structure: init precice...")
precice_tau = interface.initialize()

if (interface.is_action_required(action_write_initial_data())):
   interface.write_block_scalar_data(crossSectionLengthID, N+1, vertexIDs, crossSectionLength)
   interface.fulfilled_action(action_write_initial_data())

interface.initialize_data()

if (interface.is_read_data_available()):
   interface.read_block_scalar_data(pressureID, N+1, vertexIDs, pressure)

crossSection0 = config.crossSection0(pressure.shape[0] - 1)
pressure0 = config.p0 * np.ones_like(pressure)

while interface.is_coupling_ongoing():
   # When an implicit coupling scheme is used, checkpointing is required
    if interface.is_action_required(action_write_iteration_checkpoint()):
        interface.fulfilled_action(action_write_iteration_checkpoint())

    crossSectionLength = crossSection0 * ((pressure0 - 2.0 * config.c_mk ** 2) ** 2 / (pressure - 2.0 * config.c_mk ** 2) ** 2)

    interface.write_block_scalar_data(crossSectionLengthID, N + 1, vertexIDs, crossSectionLength)
    interface.advance(precice_tau)
    interface.read_block_scalar_data(pressureID, N + 1, vertexIDs, pressure)

    if interface.is_action_required(action_read_iteration_checkpoint()): # i.e. not yet converged
        interface.fulfilled_action(action_read_iteration_checkpoint())
    else:
        t += precice_tau

print("Exiting StructureSolver")

interface.finalize()

